daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -1818,10 +1818,62 @@ func (devices *DeviceSet) AddDevice(hash, baseHash string) error {
 		return fmt.Errorf("devmapper: device %s already exists. Deleted=%v", hash, info.Deleted)
 	}
 
-	if err := devices.createRegisterSnapDevice(hash, baseInfo); err != nil {
+	devinfo := &devInfo{}
+
+	if err := devices.parseStorageOpt(storageOpt, devinfo); err != nil {
+		return err
+	}
+
+	if devinfo.Size == 0 {
+		devinfo.Size = baseInfo.Size
+	}
+
+	if devinfo.Size < baseInfo.Size {
+		return fmt.Errorf("devmapper: Container size cannot be smaller than %s", units.HumanSize(float64(baseInfo.Size)))
+	}
+
+	if err := devices.createRegisterSnapDevice(hash, baseInfo, devinfo.Size); err != nil {
 		return err
 	}
 
+	// Grow the container rootfs.
+	if devinfo.Size > 0 {
+		info, err := devices.lookupDevice(hash)
+		if err != nil {
+			return err
+		}
+		if info == nil {
```

Comment:
no need to check for info == nil since it's already done in lookupDevice and at this point info is != nil